src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -51,11 +55,40 @@ object RuleUtils {
       }
     }
 
+    def isHybridScanCandidate(entry: IndexLogEntry, files: Seq[FileInfo]): Boolean = {
+      // TODO: Some threshold about the similarity of source data files - number of common files or
+      //  total size of common files.
+      //  See https://github.com/microsoft/hyperspace/issues/159
+      // TODO: As in [[PlanSignatureProvider]], Source plan signature comparison is required to
+      //  support arbitrary source plans at index creation.
+      //  See https://github.com/microsoft/hyperspace/issues/158
+
+      // Find a common file between the input relation & index source files.
+      // Without the threshold described above, we can utilize exists & contain functions here.
+      files.exists(entry.allSourceFileInfos.contains)
+    }
+
     // TODO: the following check only considers indexes in ACTIVE state for usage. Update
     //  the code to support indexes in transitioning states as well.
+    //  See https://github.com/microsoft/hyperspace/issues/65
     val allIndexes = indexManager.getIndexes(Seq(Constants.States.ACTIVE))
 
-    allIndexes.filter(index => index.created && signatureValid(index))
+    if (hybridScanEnabled) {
+      val files = plan
+        .collect {
+          case LogicalRelation(
+              HadoopFsRelation(location: PartitioningAwareFileIndex, _, _, _, _, _),
+              _,
+              _,
+              _) =>
+            location.allFiles.map(f =>
+              FileInfo(f.getPath.toString, f.getLen, f.getModificationTime))
+        }
+      assert(files.length == 1)
```

Comment:
Is there a better name than `files`? maybe `filesByRelations`, `filesGroupedByRelation`, etc?

Reply:
`filesByRelations` looks good :)